server/src/main/java/org/elasticsearch/common/time/JavaDateFormatter.java

```diff
@@ -122,30 +111,58 @@ public TemporalAccessor parse(String input) {
         }
 
         try {
-            return parser.parse(input);
+            return doParse(input);
         } catch (DateTimeParseException e) {
             throw new IllegalArgumentException("failed to parse date field [" + input + "] with format [" + format + "]", e);
         }
     }
 
+    /**
+     * Attempt parsing the input without throwing exception. This is needed because java-time requires ordering on optional (composite)
+     * patterns. Joda does not suffer from this.
+     * https://bugs.openjdk.java.net/browse/JDK-8188771
+     *
+     * @param input An arbitrary string resembling the string representation of a date or time
+     * @return true if parsing was successful, false if parsing failed
+     */
+    private TemporalAccessor doParse(String input) {
+        if (parsers.size() > 1) {
+            for (DateTimeFormatter formatter : parsers) {
+                ParsePosition pos = new ParsePosition(0);
+                Object object = formatter.toFormat().parseObject(input, pos);
```

Comment:
there is no need to call `formatter.toFormat()` with every invocation, this could even be done in the ctor, right?

Reply:
you mean keeping a `Collection<Format>` and refactoring all usages of `DateTimeFormatter`to be a usages of `Format` ? 
Having a field of type `Format` could be misleading that any `Format` implementation can be used. Whereas it is `DateTimeFormatter.ClassicFormat` that we depend on.
I thin it would be better to stick with DateTimeFormat across the class (we use this class anyway for `printer` and`roundUpBuilder`) and only use `Format` as a implementation details.

Reply:
I'd consider this optimization optional for now.